do not merge: Multi window mvp

lucas-koehler opened the pull request.

sdirix commented on packages/secondary-window-ui/src/browser/secondary-window-ui-frontend-contribution.ts:
```diff
@@ -0,0 +1,59 @@
+            execute: async widget => {
+
+                // sanity check
+                if (!ExtractableWidget.is(widget)) {
+                    // command executed with a non-extractable widget
+                    console.error('Invalid command execution');
+                }
```
Should we return here?

sdirix commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -358,14 +401,10 @@ export class ElectronMainApplication {
         const y = Math.round(bounds.y + (bounds.height - height) / 2);
         const x = Math.round(bounds.x + (bounds.width - width) / 2);
         return {
-            frame: this.useNativeWindowFrame,
-            isFullScreen: false,
-            isMaximized: false,
             width,
             height,
             x,
-            y,
-            ...this.getDefaultOptions()
```
Is there a consequence because of removing all these existing options?

sdirix commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -315,6 +316,38 @@ export class ElectronMainApplication {
+            };
+            if (!this.useNativeWindowFrame) {
+                // If the main window does not have a native window frame, do not show  an icon in the secondary window's native title bar.
+                // The data url is a 1x1 transparent png
```
Where is this 1x1 transparent png encoding from? Did you generate it or copy? Can this be encoded differently?

sdirix commented on packages/core/src/browser/widgets/extractable-widget.ts:
```diff
@@ -0,0 +1,33 @@
+export interface ExtractableWidget extends Widget {
+    /** Set to `true` to mark the widget to be extractable. */
+    isExtractable: boolean;
+    /** The secondary window that the window was extracted to or `undefined` if it is in the main window. */
```
```suggestion
    /** The secondary window that the window was extracted to or `undefined` if it is not yet extracted. */
```

sdirix commented on packages/core/src/browser/widgets/extractable-widget.ts:
```diff
@@ -0,0 +1,33 @@
+    /** Set to `true` to mark the widget to be extractable. */
+    isExtractable: boolean;
+    /** The secondary window that the window was extracted to or `undefined` if it is in the main window. */
+    secondaryWindow: Window | undefined;
```
What happens when a widget was extracted, the window was closed and the same view opened again in the main window? Is the `secondaryWindow` set or is it a new widget instance anyway?

sdirix commented on packages/core/src/browser/secondary-window-handler.ts:
```diff
@@ -0,0 +1,200 @@
+        newWindow.onload = () => {
+            // Use the widget's title as the window title
+            // Even if the widget's label were malicious, this should be safe against XSS because the HTML standard defines this is inserted via a text node.
+            // See https://stackoverflow.com/a/61435099 and https://html.spec.whatwg.org/multipage/dom.html#document.title
```
I would not cite stackoverflow here

sdirix commented on packages/core/src/browser/secondary-window-handler.ts:
```diff
@@ -0,0 +1,200 @@
+            return;
+        }
+        if (!widget.isExtractable) {
+            console.warn('Widget is not extractable.', widget.id);
```
Shouldn't this also be an error?

sdirix commented on packages/core/src/browser/secondary-window-handler.ts:
```diff
@@ -0,0 +1,200 @@
+                return;
+            }
+
+            // Filter setImmediate messages. Do not log because these come in with very high frequency.
```
```suggestion
            // Filter setImmediate messages. Do not forward because these come in with very high frequency.
```

Can you also explain why these are not needed besides them coming in frequently?

sdirix reviewed (changes requested):
I tested the change on Linux with the Electron and Browser example applications. Works very well for me! Great work.

These are the issues I found:

 - If the extracted widget is made active, the secondary window does not receive focus nor "moves in front", i.e. stays hidden behind other windows. Example use case is the Draw IO extension which adds an editor for `.drawio` files. When extracting such a view to the secondary window, the view becomes active whenever the user clicks on the file in the main window. The view becomes visually selected, however the second window stays hidden behind other windows and it does not actually have the focus. If that is hard to fix, I could see this as a follow up to the MVP.
 
   https://user-images.githubusercontent.com/8998368/162159995-5281db45-5f31-4be3-b318-6a188f7e489b.mp4
 - As can be seen in the video above, upon selection the secondary window in Electron shows a vertical scrollbar for a short time. As Theia handles its scrollbars itself, I would like to suggest to completely disable horizontal and verticall scrollbars for secondary windows. In the Browser case it's even worse as the vertical scrollbar is there at all times:
 
https://user-images.githubusercontent.com/8998368/162161510-7ea4b58d-1d70-43c9-a697-ae60ad278154.mp4

 - Exiting the Browser Frontend does not close secondary windows. They live on in "limbo", i.e. their messages are lost and the view is de facto disconnected. When closing a frontend instance all secondary windows should also be closed, or if that's not possible, indicate that their corresponding frontend instance is gone.

https://user-images.githubusercontent.com/8998368/162175815-7e27e01a-470f-47e8-834d-2f1844596e55.mp4

lucas-koehler commented:
@sdirix Thank you very much for the detailed review and test. I rebased onto the upstream master branch and implemented the improvements.
Only bringing up hidden secondary windows on Electron was not implemented because it does not seem to simply work via the 'normal' JS `window` API. This would need further research and maybe even electron specific handling to get hold of electron's `BrowserWindow` object for the JS window object.

sdirix reviewed (changes requested):
Looks good and works for me!

Missing for a first "official" PR is a Changelog entry.

sdirix commented on CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@
 
 - [core] Move code for untitled resources into `core` from `plugin-ext` and allow users to open untitled editors with `New File` command. [#10868](https://github.com/eclipse-theia/theia/pull/10868)
 - [plugin] added support for `SnippetString.appendChoice` [#10969](https://github.com/eclipse-theia/theia/pull/10969) - Contributed on behalf of STMicroelectronics
+- [core] Added support for moving webview-based views into a secondary window/tab [#xx](https://github.com/eclipse-theia/theia/pull/xx)- Contributed on behalf of ST Microelectronics and Ericsson and by ARM and EclipseSource
```
Should we add some information regarding the ui extension too?

koegel commented:
In the browser only: It seems to me that only the first secondary window closes when the main window is closed, see attached video also.

https://user-images.githubusercontent.com/1044916/164064127-016ed083-eb0e-498e-b1d6-6953dcf39e69.mp4

lucas-koehler commented:
Thanks for the extensive reviews and testing. I'm closing this now as the official PR was opened.
